File: Scrubimail-FE/public/sdks/javascript/scrubimail.js
/**
 * ScrubiMail JavaScript SDK
 * Official SDK for the ScrubiMail Email Validation API
 * 
 * @version 1.0.0
 * @author ScrubiMail Team
 */

class ScrubiMailSDK {
  constructor(apiKey, baseURL = 'https://api.scrubimail.com/scrubimail/api/v1') {
    if (!apiKey) {
      throw new Error('API key is required');
    }
    
    this.apiKey = apiKey;
    this.baseURL = baseURL;
    this.headers = {
      'Authorization': `Bearer ${apiKey}`,
      'Content-Type': 'application/json',
      'User-Agent': 'ScrubiMail-JS-SDK/1.0.0'
    };
  }

  /**
   * Make HTTP request to API
   * @private
   */
  async _request(endpoint, options = {}) {
    const url = `${this.baseURL}${endpoint}`;
    const config = {
      headers: this.headers,
      ...options
    };

    try {
      const response = await fetch(url, config);
      
      if (!response.ok) {
        const errorData = await response.json().catch(() => ({}));
        throw new Error(errorData.message || `HTTP ${response.status}: ${response.statusText}`);
      }

      return await response.json();
    } catch (error) {
      if (error.name === 'TypeError' && error.message.includes('fetch')) {
        throw new Error('Network error: Unable to connect to ScrubiMail API');
      }
      throw error;
    }
  }

  /**
   * Validate a single email address
   * @param {string} email - Email address to validate
   * @param {boolean} realTime - Whether to perform real-time validation
   * @returns {Promise<Object>} Validation result
   */
  async validateEmail(email, realTime = true) {
    if (!email || typeof email !== 'string') {
      throw new Error('Valid email address is required');
    }

    return this._request('/validate/', {
      method: 'POST',
      body: JSON.stringify({
        email: email.trim(),
        real_time: realTime
      })
    });
  }

  /**
   * Submit bulk email validation job
   * @param {string[]} emails - Array of email addresses to validate
   * @returns {Promise<Object>} Bulk job response
   */
  async validateBulk(emails) {
    if (!Array.isArray(emails) || emails.length === 0) {
      throw new Error('Array of emails is required');
    }

    if (emails.length > 10000) {
      throw new Error('Maximum 10,000 emails allowed per batch');
    }

    return this._request('/validate-bulk/', {
      method: 'POST',
      body: JSON.stringify({
        emails: emails.map(email => email.trim())
      })
    });
  }

  /**
   * Get bulk job status
   * @param {number} jobId - Bulk job ID
   * @returns {Promise<Object>} Job status and progress
   */
  async getBulkJobStatus(jobId) {
    if (!jobId || typeof jobId !== 'number') {
      throw new Error('Valid job ID is required');
    }

    return this._request(`/bulk-status/${jobId}/`);
  }

  /**
   * Get validation status
   * @param {number} validationId - Validation ID
   * @returns {Promise<Object>} Validation result
   */
  async getValidationStatus(validationId) {
    if (!validationId || typeof validationId !== 'number') {
      throw new Error('Valid validation ID is required');
    }

    return this._request(`/status/${validationId}/`);
  }

  /**
   * Get validation history
   * @param {Object} params - Query parameters
   * @returns {Promise<Object>} Validation history
   */
  async getValidationHistory(params = {}) {
    const queryString = new URLSearchParams(params).toString();
    const endpoint = queryString ? `/history/?${queryString}` : '/history/';
    
    return this._request(endpoint);
  }

  /**
   * Get validation analytics
   * @param {Object} params - Query parameters
   * @returns {Promise<Object>} Analytics data
   */
  async getValidationAnalytics(params = {}) {
    const queryString = new URLSearchParams(params).toString();
    const endpoint = queryString ? `/analytics/?${queryString}` : '/analytics/';
    
    return this._request(endpoint);
  }

  /**
   * Get domain reputation
   * @param {string} domain - Domain to check
   * @returns {Promise<Object>} Domain reputation data
   */
  async getDomainReputation(domain) {
    if (!domain || typeof domain !== 'string') {
      throw new Error('Valid domain is required');
    }

    return this._request(`/domain-reputation/${encodeURIComponent(domain)}/`);
  }

  /**
   * Batch validate emails with progress callback
   * @param {string[]} emails - Array of email addresses
   * @param {Function} progressCallback - Callback for progress updates
   * @param {number} pollInterval - Polling interval in milliseconds (default: 2000)
   * @returns {Promise<Object>} Final validation results
   */
  async validateBulkWithProgress(emails, progressCallback = null, pollInterval = 2000) {
    // Submit bulk job
    const jobResponse = await this.validateBulk(emails);
    const jobId = jobResponse.job_id;

    // Poll for completion
    while (true) {
      const status = await this.getBulkJobStatus(jobId);
      
      if (progressCallback && typeof progressCallback === 'function') {
        progressCallback(status);
      }

      if (status.status === 'completed' || status.status === 'failed') {
        return status;
      }

      // Wait before next poll
      await new Promise(resolve => setTimeout(resolve, pollInterval));
    }
  }
}

// Export for different module systems
if (typeof module !== 'undefined' && module.exports) {
  // Node.js
  module.exports = ScrubiMailSDK;
} else if (typeof define === 'function' && define.amd) {
  // AMD
  define([], function() {
    return ScrubiMailSDK;
  });
} else {
  // Browser global
  window.ScrubiMailSDK = ScrubiMailSDK;
}

// Usage examples:
/*
// Initialize SDK
const scrubimail = new ScrubiMailSDK('your-api-key');

// Validate single email
scrubimail.validateEmail('user@example.com')
  .then(result => console.log(result))
  .catch(error => console.error(error));

// Validate bulk emails with progress
const emails = ['user1@example.com', 'user2@example.com'];
scrubimail.validateBulkWithProgress(emails, (progress) => {
  console.log(`Progress: ${progress.progress}%`);
})
.then(results => console.log('Final results:', results))
.catch(error => console.error(error));

// Get analytics
scrubimail.getValidationAnalytics({
  start_date: '2024-01-01',
  end_date: '2024-01-31'
})
.then(analytics => console.log(analytics))
.catch(error => console.error(error));
*/

File: Scrubimail-FE/public/sdks/python/scrubimail.py
"""
ScrubiMail Python SDK
Official SDK for the ScrubiMail Email Validation API

Version: 1.0.0
Author: ScrubiMail Team
"""

import json
import time
import urllib.request
import urllib.parse
import urllib.error
from typing import Dict, List, Optional, Callable, Any


class ScrubiMailError(Exception):
    """Custom exception for ScrubiMail API errors"""
    pass


class ScrubiMailSDK:
    """
    ScrubiMail Python SDK for email validation
    
    Args:
        api_key (str): Your ScrubiMail API key
        base_url (str): API base URL (default: https://api.scrubimail.com/scrubimail/api/v1)
    """
    
    def __init__(self, api_key: str, base_url: str = "https://api.scrubimail.com/scrubimail/api/v1"):
        if not api_key:
            raise ValueError("API key is required")
        
        self.api_key = api_key
        self.base_url = base_url.rstrip('/')
        self.headers = {
            'Authorization': f'Bearer {api_key}',
            'Content-Type': 'application/json',
            'User-Agent': 'ScrubiMail-Python-SDK/1.0.0'
        }
    
    def _request(self, endpoint: str, method: str = 'GET', data: Optional[Dict] = None) -> Dict[str, Any]:
        """
        Make HTTP request to API
        
        Args:
            endpoint (str): API endpoint
            method (str): HTTP method
            data (dict): Request data
            
        Returns:
            dict: API response
            
        Raises:
            ScrubiMailError: If API request fails
        """
        url = f"{self.base_url}{endpoint}"
        
        # Prepare request
        req_data = None
        if data:
            req_data = json.dumps(data).encode('utf-8')
        
        request = urllib.request.Request(
            url,
            data=req_data,
            headers=self.headers,
            method=method
        )
        
        try:
            with urllib.request.urlopen(request) as response:
                response_data = response.read().decode('utf-8')
                return json.loads(response_data)
                
        except urllib.error.HTTPError as e:
            try:
                error_data = json.loads(e.read().decode('utf-8'))
                error_message = error_data.get('message', f'HTTP {e.code}: {e.reason}')
            except (json.JSONDecodeError, AttributeError):
                error_message = f'HTTP {e.code}: {e.reason}'
            
            raise ScrubiMailError(error_message)
            
        except urllib.error.URLError as e:
            raise ScrubiMailError(f'Network error: {e.reason}')
        
        except json.JSONDecodeError as e:
            raise ScrubiMailError(f'Invalid JSON response: {e}')
    
    def validate_email(self, email: str, real_time: bool = True) -> Dict[str, Any]:
        """
        Validate a single email address
        
        Args:
            email (str): Email address to validate
            real_time (bool): Whether to perform real-time validation
            
        Returns:
            dict: Validation result
            
        Raises:
            ScrubiMailError: If validation fails
        """
        if not email or not isinstance(email, str):
            raise ValueError("Valid email address is required")
        
        data = {
            'email': email.strip(),
            'real_time': real_time
        }
        
        return self._request('/validate/', 'POST', data)
    
    def validate_bulk(self, emails: List[str]) -> Dict[str, Any]:
        """
        Submit bulk email validation job
        
        Args:
            emails (list): List of email addresses to validate
            
        Returns:
            dict: Bulk job response
            
        Raises:
            ScrubiMailError: If submission fails
        """
        if not isinstance(emails, list) or len(emails) == 0:
            raise ValueError("List of emails is required")
        
        if len(emails) > 10000:
            raise ValueError("Maximum 10,000 emails allowed per batch")
        
        data = {
            'emails': [email.strip() for email in emails]
        }
        
        return self._request('/validate-bulk/', 'POST', data)
    
    def get_bulk_job_status(self, job_id: int) -> Dict[str, Any]:
        """
        Get bulk job status
        
        Args:
            job_id (int): Bulk job ID
            
        Returns:
            dict: Job status and progress
            
        Raises:
            ScrubiMailError: If request fails
        """
        if not isinstance(job_id, int) or job_id <= 0:
            raise ValueError("Valid job ID is required")
        
        return self._request(f'/bulk-status/{job_id}/')
    
    def get_validation_status(self, validation_id: int) -> Dict[str, Any]:
        """
        Get validation status
        
        Args:
            validation_id (int): Validation ID
            
        Returns:
            dict: Validation result
            
        Raises:
            ScrubiMailError: If request fails
        """
        if not isinstance(validation_id, int) or validation_id <= 0:
            raise ValueError("Valid validation ID is required")
        
        return self._request(f'/status/{validation_id}/')
    
    def get_validation_history(self, **params) -> Dict[str, Any]:
        """
        Get validation history
        
        Args:
            **params: Query parameters (page, page_size, start_date, end_date, status)
            
        Returns:
            dict: Validation history
            
        Raises:
            ScrubiMailError: If request fails
        """
        if params:
            query_string = urllib.parse.urlencode(params)
            endpoint = f'/history/?{query_string}'
        else:
            endpoint = '/history/'
        
        return self._request(endpoint)
    
    def get_validation_analytics(self, **params) -> Dict[str, Any]:
        """
        Get validation analytics
        
        Args:
            **params: Query parameters (start_date, end_date)
            
        Returns:
            dict: Analytics data
            
        Raises:
            ScrubiMailError: If request fails
        """
        if params:
            query_string = urllib.parse.urlencode(params)
            endpoint = f'/analytics/?{query_string}'
        else:
            endpoint = '/analytics/'
        
        return self._request(endpoint)
    
    def get_domain_reputation(self, domain: str) -> Dict[str, Any]:
        """
        Get domain reputation
        
        Args:
            domain (str): Domain to check
            
        Returns:
            dict: Domain reputation data
            
        Raises:
            ScrubiMailError: If request fails
        """
        if not domain or not isinstance(domain, str):
            raise ValueError("Valid domain is required")
        
        encoded_domain = urllib.parse.quote(domain, safe='')
        return self._request(f'/domain-reputation/{encoded_domain}/')
    
    def validate_bulk_with_progress(self, 
                                   emails: List[str], 
                                   progress_callback: Optional[Callable] = None,
                                   poll_interval: int = 2) -> Dict[str, Any]:
        """
        Batch validate emails with progress callback
        
        Args:
            emails (list): List of email addresses
            progress_callback (callable): Callback for progress updates
            poll_interval (int): Polling interval in seconds
            
        Returns:
            dict: Final validation results
            
        Raises:
            ScrubiMailError: If validation fails
        """
        # Submit bulk job
        job_response = self.validate_bulk(emails)
        job_id = job_response['job_id']
        
        # Poll for completion
        while True:
            status = self.get_bulk_job_status(job_id)
            
            if progress_callback and callable(progress_callback):
                progress_callback(status)
            
            if status['status'] in ['completed', 'failed']:
                return status
            
            # Wait before next poll
            time.sleep(poll_interval)


# Usage examples:
if __name__ == "__main__":
    # Initialize SDK
    sdk = ScrubiMailSDK('your-api-key')
    
    try:
        # Validate single email
        result = sdk.validate_email('user@example.com')
        print("Single validation result:", result)
        
        # Validate bulk emails with progress
        emails = ['user1@example.com', 'user2@example.com']
        
        def progress_callback(status):
            print(f"Progress: {status.get('progress', 0)}%")
        
        bulk_result = sdk.validate_bulk_with_progress(emails, progress_callback)
        print("Bulk validation result:", bulk_result)
        
        # Get analytics
        analytics = sdk.get_validation_analytics(
            start_date='2024-01-01',
            end_date='2024-01-31'
        )
        print("Analytics:", analytics)
        
    except ScrubiMailError as e:
        print(f"ScrubiMail API Error: {e}")
    except ValueError as e:
        print(f"Validation Error: {e}")
    except Exception as e:
        print(f"Unexpected Error: {e}")